### PR TITLE
Dont do async work on process.on('exit') 

### LINKED ISSUE #47180: Extension telemetry doesn't get reported from the CLI
The `extensionGallery:install` and related events don't seem to be reported when installing extensions from the CLI.

After a quick code review, they seem to be emitted from the ExtensionsWorkbenchService. Should we move them to the ExtensionManagementService instead?

cc @seanmcbreen

### PR DESCRIPTION
**Pull request by ramya-rao-a**
From https://nodejs.org/api/process.html#process_event_exit about use of `process.on('exit')`

> Listener functions must only perform synchronous operations. The Node.js process will exit immediately after calling the 'exit' event listeners causing any additional work still queued in the event loop to be abandoned

This PR fixes #47180

All pending telemetry events get flushed out when appInsightsAppender is disposed. Since the dispose call was made from `process.once('exit')`, the appender never got disposed and so the events never really got fired.

### CODE REVIEW

File: src/vs/code/node/cliProcessMain.ts [hunk 1]
```diff
@@ -246,7 +242,10 @@ export function main(argv: ParsedArgs): TPromise<void> {
 			const instantiationService2 = instantiationService.createChild(services);
 			const main = instantiationService2.createInstance(Main);
 
-			return main.run(argv);
+			return (main.run(argv) || TPromise.as(null)).then(() => {
```
Comment (joaomoreno) [inline src/vs/code/node/cliProcessMain.ts]: Might just be cleaner to enforce that `run` always returns a promise than doing this `||`. I think this is the only place where we don't return a promise:

https://github.com/Microsoft/vscode/blob/master/src/vs/code/node/cliProcessMain.ts#L80

Comment (joaomoreno) [inline src/vs/code/node/cliProcessMain.ts]: Might just be cleaner to enforce that `run` always returns a promise than doing this `||`. I think this is the only place where we don't return a promise:

https://github.com/Microsoft/vscode/blob/master/src/vs/code/node/cliProcessMain.ts#L80

File: src/vs/platform/telemetry/common/telemetryUtils.ts [hunk 2]
```diff
@@ -30,13 +30,17 @@ export const NullTelemetryService = new class implements ITelemetryService {
...
 export function combinedAppender(...appenders: ITelemetryAppender[]): ITelemetryAppender {
-	return { log: (e, d) => appenders.forEach(a => a.log(e, d)) };
+	return {
+		log: (e, d) => appenders.forEach(a => a.log(e, d)),
+		dispose: () => TPromise.join(appenders.map(a => a.dispose()))
+	};
 }
 
-export const NullAppender: ITelemetryAppender = { log: () => null };
+export const NullAppender: ITelemetryAppender = { log: () => null, dispose: () => null };
```
Comment (joaomoreno) [inline src/vs/platform/telemetry/common/telemetryUtils.ts]: It should return `TPromise.as(null)` instead of `null`.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
This makes sense, does the data get flushed now? Thanks for the investigation.

File: src/vs/code/node/cliProcessMain.ts (hunk 1 above)
Comment (ramya-rao-a) [reply src/vs/code/node/cliProcessMain.ts]: Done

File: src/vs/platform/telemetry/common/telemetryUtils.ts (hunk 2 above)
Comment (ramya-rao-a) [reply src/vs/platform/telemetry/common/telemetryUtils.ts]: Done

**Comment by ramya-rao-a** [comment]
Yes, with these changes the telemetry data is being flushed.

**Review by joaomoreno — APPROVED** [review]

**Pull request merged**